### PR TITLE
7 adding testing framework

### LINKED ISSUE #7: Implement Backend Unit Testing
Currently the back-end code lacks unit testing.  This ticket is responsible for setting up the initial unit testing framework and is required before any unit tests are implemented.

### PR DESCRIPTION
**Pull request by DarrellRichards**
# Pull Request

Added support for the Jest testing framework, along with setting up support for coverage reports (Only works once there is a single test available). 

Just want to make a note since there are no tests written on the jest side of things. When running the test command it will fail with an exit code of 1. 

If you wish to test so that it does not fail with an exit code of 1 run the following command: `npm run test:node --passWithNoTests`

### Requirements Added
- Jest
- Babel -> added for support for Jest with typescript
- ts-jest -> added for support for Jest with typescript

Fixes #7 

## Type of change

- [x] A new feature
- [ ] A bug fix
- [ ] Documentation only changes
- [x] Changes that do not affect the meaning of the code (white-space, formatting, missing semi-colons, etc)
- [x] A code change that neither fixes a bug nor adds a feature
- [ ] A code change that improves performance
- [ ] Adding missing tests or correcting existing tests
- [ ] Changes that affect the build system or external dependencies
- [ ] Changes to our CI configuration files and scripts
- [ ] Other changes that don't modify src or test files

Add description here

```release-note
   - Added Jest framework as the test runner for the node src
```

---

## Checklist

- [x] My code follows the [contributing](https://github.com/softrams/bulwark/blob/master/CONTRIBUTING.md) guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

**Comment by Whamo12** [comment]
@DarrellRichards Could you open this pull request against the Develop branch?

**Comment by DarrellRichards** [comment]
@Whamo12 

My apologizes, I see it wasn't defaulted to `develop`.

I should have the last couple of changes up here in a moment.

File: CONTRIBUTING.md [hunk 1]
```diff
@@ -42,3 +42,16 @@ Any release from master will have a unique version
 ```
 npm run lint
 ```
+
+### Testing
```
Comment (DarrellRichards) [inline CONTRIBUTING.md:46]: Not sure if this is the best place to document the testing changes I saw this where the `linting` documented so I thought `testing` would fit in. I can move if needed

File: jest.config.js [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+  roots: ['<rootDir>/src'],
+  transform: {
+    '^.+\\.ts?$': 'ts-jest',
+  },
+  testRegex: '(/__tests__/.*|(\\.|/)(test|spec))\\.ts?$',
+  collectCoverage: true,
+  collectCoverageFrom: [
+    "src/**/*.{ts,tsx}"
+  ],
+  moduleFileExtensions: ['ts', 'tsx', 'js', 'jsx', 'json', 'node'],
```
Comment (Whamo12) [inline jest.config.js:19]: We do not need `tsx` or `jsx`.

File: jest.config.js [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+  // The directory where Jest should output its coverage files
+  coverageDirectory: "coverage",
+  roots: ['<rootDir>/src'],
+  transform: {
+    '^.+\\.ts?$': 'ts-jest',
+  },
+  testRegex: '(/__tests__/.*|(\\.|/)(test|spec))\\.ts?$',
+  collectCoverage: true,
+  collectCoverageFrom: [
+    "src/**/*.{ts,tsx}"
```
Comment (Whamo12) [inline jest.config.js:17]: Remove `tsx`

File: CONTRIBUTING.md (hunk 1 above)
Comment (Whamo12) [reply CONTRIBUTING.md:46]: This is fine for now.

**Comment by DarrellRichards** [comment]
@Whamo12

I pushed the updates to remove `tsx and jsx` (I wasn't sure if there was plans in the future).

Also, the tests fail with the command till we have a test wrote.

**Pull request merged**